text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,228 @@
+    /// return its final result, otherwise return the value
+    /// that will get passed to `logical_or()` (normally this
+    /// means returning self back, but you can change the value).
+    fn short_circuit_or(self) -> ShortCircuit<Self::Output, Intermediate>;
+
+    /// Complete the *logical or* in case it did not short-circuit.
+    /// Normally this would just return `rhs`.
+    fn logical_or(lhs: Intermediate, rhs: Rhs) -> Self::Output;
+}
+
+trait LogicalAnd<Rhs = Self>: Sized {
+    type Output;
+    type Intermediate;
+
+    /// Decide whether the *logical and* should short-circuit
+    /// or not based on its left-hand side argument. If so,
+    /// return its final result, otherwise return the value
+    /// that will get passed to `logical_and()` (normally this
+    /// means returning self back, but you can change the value).
+    fn short_circuit_and(self) -> ShortCircuit<Self::Output, Intermediate>;
+
+    /// Complete the *logical and* in case it did not short-circuit.
+    fn logical_and(lhs: Intermediate, rhs: Rhs) -> Self::Output;
+}
+```
+
+With a matching desugaring:
+
+```rust
+<expr_a::<T>> || <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_or() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_or(lhs, expr_b)
+}
+```
+
+and
+
+```rust
+<expr_a::<T>> && <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_and() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_and(lhs, expr_b)
+}
+```
+
+From taking into consideration the current functions, and previous discussion on [internals](https://internals.rust-lang.org/t/pre-rfc-overload-short-curcuits/10460) it seems that the following makes the most sense in terms of outcomes.
+
+#### For `Option<T>`:
+
+```rust
+fn foo() -> Option<i32> {
+    Some(3)
+}
+
+fn main() {
+    Some(4) || Some(5); // == Some(4)
+    None || Some(5);    // == Some(5)
+    Some(4) || foo();   // == Some(4) (foo is *not* called)
+    None || foo();      // == Some(3) (foo is called)
+    None || 3;          // == 3
+    Some(2) || 1;       // == 2
+    Some(1) || panic!() // == Some(1)               + These two are side effects from !
+    None || return      // returns from function    + and are the same to how boolean || works
+    Some(2) && Some(3)  // Some(3)
+    None && Some(1)     // None
+    Some(3) && None     // None
+
+    Some(2) || Some("hello") // Error: LogicalOr<Option<&str>> not implemented for Option<i32>
+    Some(2) || 2 || 3        // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+#### For `Result<T, E>`
+```rust
+struct MyError;
+
+fn foo() -> Result<i32, MyError> {
+    Ok(3)
+}
+
+fn main() {
+    Ok(4) || Ok(5);           // == Ok(4)
+    Err(MyError) || Ok(5);  // == Ok(5)
+    Ok(4) || foo();           // == Ok(4) (foo is *not* called)
+    Err(MyError) || foo();  // == Ok(3) (foo is called)
+    Err(MyError) || 3;      // == 3
+    Ok(2) || 1;               // == 2
```

Comment:
```suggestion
    // or
    Ok(4) || Ok(5);           // == Ok(4)
    Err(MyError) || Ok(5);  // == Ok(5)
    // or_else (without taking parameter)
    Ok(4) || foo();           // == Ok(4) (foo is *not* called)
    Err(MyError) || foo();  // == Ok(3) (foo is called)
    // unwrap_or
    Err(MyError) || 3;      // == 3
    Ok(2) || 1;               // == 2
    // unwrap_or_else (without taking parameter)
    Ok(4) || bar();           // == 4 (foo is *not* called)
    Err(MyError) || bar();  // == 3 (foo is called)
```

`bar()` returns `i32`, but I think this is kinda hard to distinguish if `xxx` in `|| xxx()` is returning what in the first place. It do two tasks which are either `unwrap_` or `or_`. Looks like a potential place for footguns.